test/test_autograd.py

```diff
@@ -347,6 +347,37 @@ def test_custom_function_exception(self):
         with self.assertRaisesRegex(Exception, "Simulate error on backward pass"):
             t3.sum().backward()
 
+    def test_custom_function_raises_if_input_returned_as_is_and_saved(self):
+        for save_inputs, mark_dirty in product(*([(True, False)] * 2)):
+            class Func(torch.autograd.Function):
+                @staticmethod
+                def forward(x):
+                    return x ** 3, x
+
+                @staticmethod
+                def setup_context(ctx, inputs, outputs):
+                    if mark_dirty:
+                        ctx.mark_dirty(inputs[0])
+
+                    if save_inputs:
+                        ctx.save_for_backward(inputs[0])
+                    else:
+                        ctx.save_for_backward(outputs[1])
+
+                @staticmethod
+                def backward(ctx, grad_output, grad_x):
+                    pass
+
+            a = torch.tensor(1., requires_grad=True).clone()
+
+            if mark_dirty:
+                _unused, b = Func.apply(a)
+                if mark_dirty:
+                    self.assertTrue(a is b)
+            else:
+                with self.assertRaisesRegex(RuntimeError, "A input that has been returned as-is"):
+                    Func.apply(a)
+
```

Comment:
- As a sanity check, does it make sense to add a test where we return an input as-is but do not save it for backward? I would expect this to not trigger the error.
- Could we add the same test, but run it under functorch.grad? It would be good to test that we're consistent between functorch and regular autograd.

Reply:
I think we already have the test in test_autograd, but sounds good on the functorch side.